### ISSUE TITLE
"Enlist in Aspire" should escape resource name in AddProject

### ISSUE THREAD

**Issue by JamesNK** [dotnet/aspire#2102]
Resource names have validation on allowed characters: ASCII alpha numeric + dashes. This is to ensure generated names are compatible with cloud provider naming and special characters used in Aspire when building connection strings.

A problem is a .NET project can contain characters that are invalid. When Aspire tooling adds an `AddProject<T>()` line to the Aspire host, it could produce an app that doesn't run. See https://github.com/dotnet/aspire/issues/2083.

The resource name in the generated `AddProject` call needs to be escaped to make it compatible.

For example, given the project name `WebAppMvc_Anc80_NoAuth_1`, the invalid underscores should be replaced with dashes:

```diff
-builder.AddProject<Projects.WebAppMvc_Anc80_NoAuth_1>("webappmvc_anc80_noauth_1");
+builder.AddProject<Projects.WebAppMvc_Anc80_NoAuth_1>("webappmvc-anc80-noauth-1");
```

**Comment by mitchdenny** [comment]
Need to test that ACA can accept dashes, and if not figure out how this impacts service discovery and how to workaround the issue.

**Comment by JamesNK** [comment]
Here are container name rules: https://learn.microsoft.com/en-us/rest/api/storageservices/naming-and-referencing-containers--blobs--and-metadata#container-names

I think we almost conform to those rules. We allow 64 characters instead of 63, and don't have a min length.

Should double check with ACA that those rules are used by them.

**Comment by davidfowl** [comment]
Can we require lowercase names ?

**Comment by davidfowl** [comment]
We also can’t hardcode ACA rules but having rules are fine

**Comment by JamesNK** [comment]
> Can we require lowercase names ?

Names are case insensitive. That should mean that ACA or another tool can make them lowercase and things still work, i.e. MYPROJECT and myproject aren't allowed so making names lowercase isn't breaking

**Comment by balachir** [comment]
I've opened bug 1976045 to track the tooling change in AzDO.

**Comment by timheuer** [comment]
Another plus for some analyzers just in case ;-) (of course our tools generating should be better though also)

**Comment by davidfowl** [comment]
@JamesNK We wanted to fix this for p4 after reverting the change in p3, did this happen?

**Comment by JamesNK** [comment]
Not yet.

**Comment by davidfowl** [comment]
Do we have to revert this change again in p4?

**Comment by JamesNK** [comment]
The VS issue is marked as a blocker for the release. Hopefully not.

**Comment by davidfowl** [comment]
@JamesNK  assigned to you, but I think @phenning is working on it

**Comment by balachir** [comment]
This issue 1976045 is fixed now in the latest builds of VS 17.10 P2 and I've validated the fix for _ and . characters in the project name.
